hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +53,130 @@
  */
 public class ManagementCenterService {
 
+    static class MCEventStore {
+        static final long MC_EVENTS_WINDOW_MILLIS = TimeUnit.SECONDS.toMillis(30);
+        private final LongSupplier clock;
+        private volatile long mostRecentAccessTimestamp;
+        private final ConcurrentMap<Address, Long> lastAccessTimestamps = new ConcurrentHashMap<>();
+        private final BlockingQueue<Event> mcEvents;
+
+        MCEventStore(LongSupplier clock, BlockingQueue<Event> mcEvents) {
+            this.clock = clock;
+            this.mostRecentAccessTimestamp = clock.getAsLong();
+            this.mcEvents = mcEvents;
+        }
+
+        void log(Event event) {
+            if (clock.getAsLong() - mostRecentAccessTimestamp > MC_EVENTS_WINDOW_MILLIS) {
+                // ignore event and clear the queue if the last poll happened a while ago
+                onMCEventWindowExceeded();
+            } else {
+                mcEvents.offer(event);
+            }
+
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
+            lastAccessTimestamps.clear();
+        }
+
+        /**
+         * @param mcRemoteAddr the address of the calling MC instance.
+         * @return the events which were added to the queue since this MC last polled, or all known events if the MC polls for the
+         * first time.
+         */
+        public List<Event> pollMCEvents(Address mcRemoteAddr) {
+            Long lastAccessObj = lastAccessTimestamps.get(mcRemoteAddr);
+            List<Event> events;
+            if (lastAccessObj == null) {
+                events = new ArrayList<>(mcEvents);
+            } else {
+                long lastAccess = lastAccessObj;
+                events = new ArrayList<>();
+                for (Event evt : mcEvents) {
+                    if (evt.getTimestamp() >= lastAccess) {
```

Comment:
What happens in the following scenario?

There is a single MC instance connected to the cluster. The events are:
```
 | time
\ /
timestamp1 event1 <- event is logged; MC polls events and timestamp1 is stored in lastAccessTimestamps
timestamp1 event2 <- event is logged
timestamp2 event3 <- event is logged; MC polls events once again
```

If I'm not mistaken, with the proposed implementation MC may receive `event1` twice. WDYT?

Also, `Clock::currentTimeMillis` is not monotonic, so there may be various anomalies when the clock is updated on the machine.

Reply:
Thanks, I changed it back for `nanoTime()`.

Reply:
> Thanks, I changed it back for `nanoTime()`.

Are you sure that it's an option? `Event#getTimestamp()` returns a wall clock timestamp as we need it to show it to users in MC UI, while with `nanoTime()` both `mostRecentAccessTimestamp` and `lastAccessTimestamps` contain monotonic clock timestamps. We can't compare them with each other, so `currentTimeMillis()` looks like a must have with the proposed design.

The issue described in my original comment is an edge case, so I'm not sure if we need to do something with that. Theoretically, we could implement a `ConcurrentArrayRingbuffer`-based polling like it's done for metrics blobs (see `MetricsService#readMetrics()` and related code).

But with this approach MC instances will have to track their sequence numbers in the ring buffer and pass them when polling events. This means some backwards compatibility problems on the cluster side as members will have to support both old and new MC versions. We could maybe solve it if members would keep a shared sequence number field to be used for old versions of MC (i.e. MC doesn't send the sequence number, the member assumes that it's an old version and uses value from this field).

WDYT?

Reply:
> so currentTimeMillis() looks like a must have with the proposed design.

aah right that's why I chose `System.currentTimeMillis()` originally. Let me get back to here a bit later.

Reply:
I pushed a commit switching back from nanos to millis.

Back to your original question: yes, if
 - an event arrives 
 - and MC polls the events in the same millisec,
then the event will be sent to MC in two polls, because `evt.getTimestamp() >= lastAccess` condition will be true in both calls.

I also considered using a strictly-greater condition (`evt.getTimestamp() > lastAccess`) in which case if
 - MC polls the events
 - and an event arrives in the same millisec,
then the event doesn't get reported at all.

So I considered the chance of duplicating events a better (and good-enough :) ) option than omitting some events.

One more option to improve the current implementation (just an idea) can be: when sending back the events, tracking how many same-millisec events were sent in the response, and excluding those in the next poll. Let me give it a go in the next comit...